Improves error message for invalid dump parameter

DanielYankura opened the pull request:
-Checks if dump parameter is valid before trying to use it. 
-If invalid it prints an error message that includes the parameter name

closes #26714

Commit:
Improves error message for invalid dump parameter

-Checks if dump parameter is valid before trying to use it.
-If invalid it prints an error message that inclues the parameter name

closes #26714

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2420989/) on b9c185d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28515/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28515/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 97af19c98a76c47db2b4157c4e73c28f68f45595`

Commit:
fixed formatting issues

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2421172/) on 04012fb wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28515/site)

This comment will be updated on new commits.

GiudGiud reviewed (commented):
looks good you ll just want to make a small exception test.
maybe in test/tests/misc?

Commit:
Added basic test for --dump paramters

-provides --dump with a bad parameter to ensure it gives the right error message

DanielYankura commented:
I added a small test case for it in test/tests/misc/check_error. If there are more test cases, or if it should be added in a different directory I can go ahead and fix that.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2422845/) on a096b81 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>97af19</th>
      <th colspan=4>#28515 a096b8</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.03%</td>
      <td>85.03%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>105193</td>
      <td>105196</td>
      <td>+3</td>
      <td>8</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18527</td>
      <td>18527</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28515/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28515/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28515/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on framework/src/base/MooseApp.C:
```diff
@@ -820,13 +820,19 @@ MooseApp::setupOptions()
+      _ready_to_exit = true;
+    }
+    else
+      mooseError("Invalid search parameter ", param_search);
```
```suggestion
      mooseError("Search parameter '", param_search, "' was not found in the registered syntax.");
```

GiudGiud commented on framework/src/base/MooseApp.C:
```diff
@@ -820,13 +820,19 @@ MooseApp::setupOptions()
-    // Turn off live printing so that it doesn't mess with the dump
-    _perf_graph.disableLivePrint();
+    // Check if second arg is valid or not
+    if ((tree.getRoot()).is_object())
```
when it is not valid, what is it? is it empty? Could we use that?

I m wondering if "is_object" might bite us later, when a valid search returned an array rather than an object for example.

DanielYankura replied:
So it's not valid when the search parameter is not found in all the info --dump prints out. Having an empty param is valid because then it just prints out everything and calling is_object() on an empty parameter returns true. I used is_object() because that's what is called later on where the error ends up getting thrown. I believe it's called in json.h. I don't believe --dump can take more than one search parameter so we shouldn't have to worry about that right now, but if that does ever get changed then I will have to change how it checks for valid search parameters.

brandonlangley replied:
An perhaps safer alternative to guard against non-objects could simply be:

```cpp
if (tree.getRoot().is_null())
  mooseError("Search parameter '", param_search, "' was not found in the registered syntax.");
```

DanielYankura replied:
From my limited testing using .is_null() instead of .is_object() should work as well! The only potential issue I see is that later on .is_object() is still called. The ambiguous error was thrown from json.h from lines 21443 to 21454.

`    const_reference operator[](const typename object_t::key_type& key) const
    {
        // const operator[] only works for objects
        if (JSON_HEDLEY_LIKELY(is_object()))
        {
            auto it = m_data.m_value.object->find(key);
            JSON_ASSERT(it != m_data.m_value.object->end());
            return it->second;
        }


        JSON_THROW(type_error::create(305, detail::concat("cannot use operator[] with a string argument with ", type_name()), this));
    }
`
I just moved the is_object() call higher up so that I could return a clearer error. I didn't want to modify this section of code because I wasn't sure where else it might be used. But if there are any cases where is_null() and is_object() could both return false then we'd end up with the same error as before. If it's impossible for them to both return false then it shouldn't matter, especially since right now --dump can only handle one search string.

GiudGiud reviewed (approved):
thanks for the fix

someone actually hit that problem on discussions last weekend
